### PR TITLE
Torch output shape support

### PR DESCRIPTION
**Pull request by ShreyaR**
This PR adds support for output and input shapes for all Ludwig components.

- Adds a `LudwigComponent` class that supports getting the `output_shape` via running `forward` on a dummy tensor of `input_shape`
- Adds `LudwigComponent` as the parent class of all modules.
- Adds `LudwigComponent` to encoders.
- Adds `LudwigComponent` to combiners.
- Update features to use `output_shape`, `input_shape` properties instead of methods `get_output_shape`, `get_input_shape`.

### CODE REVIEW

File: ludwig/utils/torch_utils.py [hunk 1]
```diff
@@ -93,6 +93,23 @@ def losses(self):
...
+
+    @property
+    def input_shape(self) -> torch.Size:
+        """ Returns the size of the input tensor without the batch dimension. """
+        raise NotImplementedError('Abstract class.')
+
+    @property
+    def output_shape(self) -> torch.Size:
+        """ Returns the size of the output tensor without the batch dimension."""
+        output_tensor = self.forward(torch.rand(*self.input_shape))
```
Comment (tgaddair) [inline ludwig/utils/torch_utils.py:109]: General advice is to avoid calling `forward` directly. Not sure it applies here, but probably good practice to follow:

https://discuss.pytorch.org/t/any-different-between-model-input-and-model-forward-input/3690

File: ludwig/utils/torch_utils.py [hunk 2]
```diff
@@ -93,6 +93,23 @@ def losses(self):
...
+        super().__init__()
+
+    @property
+    def input_shape(self) -> torch.Size:
+        """ Returns the size of the input tensor without the batch dimension. """
+        raise NotImplementedError('Abstract class.')
+
+    @property
+    def output_shape(self) -> torch.Size:
+        """ Returns the size of the output tensor without the batch dimension."""
```
Comment (tgaddair) [inline ludwig/utils/torch_utils.py:108]: The input shape will not include batch size, so we should account for that. PyTorch Summary uses batch size 2 for batch norm. Here's an example:

https://github.com/sksq96/pytorch-summary/blob/master/torchsummary/torchsummary.py#L59

File: ludwig/utils/torch_utils.py [hunk 3]
```diff
@@ -93,6 +93,23 @@ def losses(self):
             collected_losses.extend(layer.losses)
         return collected_losses
 
+
+class LudwigComponent(Module):
```
Comment (tgaddair) [inline ludwig/utils/torch_utils.py:97]: We probably want to inherit from `LudwigModule` so we can get the loss tracking features.

File: ludwig/features/binary_feature.py [hunk 4]
```diff
@@ -232,8 +232,9 @@ def get_prediction_set(self):
     def get_output_dtype(cls):
         return tf.bool
 
-    def get_output_shape(self):
-        return ()
+    @property
+    def output_shape(self) -> torch.Size:
```
Comment (tgaddair) [inline ludwig/features/binary_feature.py:236]: This should be `torch.Size([1])`

File: ludwig/features/binary_feature.py [hunk 5]
```diff
@@ -145,8 +145,8 @@ def call(self, inputs, training=None, mask=None):
     def get_input_dtype(cls):
         return tf.bool
 
-    def get_input_shape(self):
-        return ()
+    # def get_input_shape(self):
+    #     return ()
```
Comment (tgaddair) [inline ludwig/features/binary_feature.py:149]: This should be `torch.Size([1])`

File: ludwig/features/h3_feature.py [hunk 6]
```diff
@@ -117,8 +118,12 @@ def call(self, inputs, training=None, mask=None):
...
         return tf.uint8
 
-    def get_input_shape(self):
-        return H3_VECTOR_LENGTH,
+    # def get_input_shape(self):
+    #     return H3_VECTOR_LENGTH,
+
+    @property
+    def input_shape(self) -> torch.Size:
+        return torch.Size((H3_VECTOR_LENGTH, ))
```
Comment (tgaddair) [inline ludwig/features/h3_feature.py:126]: For consistency, I would use a list instead of a tuple here.

File: ludwig/utils/torch_utils.py [hunk 7]
```diff
@@ -93,6 +93,23 @@ def losses(self):
...
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def input_shape(self) -> torch.Size:
```
[the remark this answers is not in the document]
Comment (ShreyaR) [reply ludwig/utils/torch_utils.py:102]: Good catch, done!

File: ludwig/features/sequence_feature.py [hunk 8]
```diff
@@ -301,8 +302,9 @@ def get_prediction_set(self):
     def get_output_dtype(cls):
         return tf.int32
 
-    def get_output_shape(self):
-        return self.max_sequence_length,
+    @property
+    def output_shape(self) -> torch.Size:
+        return torch.Size((self.max_sequence_length, ))
```
Comment (tgaddair) [inline ludwig/features/sequence_feature.py:307]: As above, we should stick to either list or tuple.

File: ludwig/features/set_feature.py [hunk 9]
```diff
@@ -136,8 +137,12 @@ def call(self, inputs, training=None, mask=None):
...
         return tf.bool
 
-    def get_input_shape(self):
-        return len(self.vocab),
+    # def get_input_shape(self):
+    #     return len(self.vocab),
+
+    @property
+    def input_shape(self) -> torch.Size:
+        return torch.Size((len(self.vocab), ))
```
Comment (tgaddair) [inline ludwig/features/set_feature.py:145]: Same as above.

File: ludwig/features/set_feature.py [hunk 10]
```diff
@@ -231,8 +236,12 @@ def get_prediction_set(self):
...
         return tf.bool
 
-    def get_output_shape(self):
-        return self.num_classes,
+    # def get_output_shape(self):
+    #     return self.num_classes,
+
+    @property
+    def output_shape(self) -> torch.Size:
+        return torch.Size((self.num_classes, ))
```
Comment (tgaddair) [inline ludwig/features/set_feature.py:244]: Same as above.

File: ludwig/utils/torch_utils.py (hunk 1 above)
Comment (ShreyaR) [reply ludwig/utils/torch_utils.py:109]: I think the advantage of using `__call__` instead of `forward` is to call the forward and backward hooks. In our case it's probably more efficient to not call the hooks since we only want to use forward to compute the shapes, if in case the forward and backward hooks end up having some expensive computation.

I'm happy to switch to the `__call__` method if you prefer though.

File: ludwig/utils/torch_utils.py (hunk 3 above)
Comment (ShreyaR) [reply ludwig/utils/torch_utils.py:97]: Done!

**Review by tgaddair — COMMENTED** [review]
In general, we should remove the commented out code and align on using either list or tuple as the param type to Size. Other than that, looks great!

File: ludwig/utils/torch_utils.py (hunk 2 above)
Comment (ShreyaR) [reply ludwig/utils/torch_utils.py:108]: Good catch, I've added a dummy batch dimension of 2 right now!

File: ludwig/utils/torch_utils.py (hunk 1 above)
Comment (tgaddair) [reply ludwig/utils/torch_utils.py:109]: Since in most cases the hooks are just side effects, we probably don't want to call them here, I agree. So for now I think we can stick with forward.

File: ludwig/features/binary_feature.py (hunk 4 above)
Comment (ShreyaR) [reply ludwig/features/binary_feature.py:237]: Done

File: ludwig/features/binary_feature.py (hunk 5 above)
Comment (ShreyaR) [reply ludwig/features/binary_feature.py]: Done

File: ludwig/features/h3_feature.py (hunk 6 above)
Comment (ShreyaR) [reply ludwig/features/h3_feature.py]: Done

File: ludwig/features/sequence_feature.py (hunk 8 above)
Comment (ShreyaR) [reply ludwig/features/sequence_feature.py]: Done

File: ludwig/features/set_feature.py (hunk 10 above)
Comment (ShreyaR) [reply ludwig/features/set_feature.py]: Done

**Review by tgaddair — APPROVED** [review]
